xgcm/test/test_transform.py

```diff
@@ -0,0 +1,744 @@
+
+    source_dim = _parse_dim(source.data)
+    target_dim = _parse_dim(target)
+
+    # parse the target_data manually
+    target_data = transform_kwargs.pop("target_data", None)
+    if target_data is None:
+        target_data = source[source_dim]
+
+    if error_flag:
+        with pytest.xfail():
+            interpolated = linear_interpolation(
+                source.data,
+                target_data,
+                target,
+                source_dim,
+                source_dim,
+                target_dim,
+                **transform_kwargs
+            )
+    else:
+        interpolated = linear_interpolation(
+            source.data,
+            target_data,
+            target,
+            source_dim,
+            source_dim,
+            target_dim,
+            **transform_kwargs
+        )
+        xr.testing.assert_allclose(interpolated, expected.data)
+
+
+def test_mid_level_conservative(conservative_cases):
+    """Test the conservative interpolations on the xarray wrapper level"""
+    (
+        source,
+        grid_kwargs,
+        target,
+        transform_kwargs,
+        expected,
+        error_flag,
+    ) = conservative_cases
+
+    # method keyword is only for high level tests
+    transform_kwargs.pop("method")
+
+    source_dim = grid_kwargs["coords"]["Z"]["center"]
+    bounds_dim = grid_kwargs["coords"]["Z"]["outer"]
+    target_dim = _parse_dim(target)
+
+    # parse the target_data manually
+    target_data = transform_kwargs.pop("target_data", None)
+    if target_data is None:
+        target_data = source[bounds_dim]
+    if error_flag:
+        with pytest.xfail():
+            interpolated = conservative_interpolation(
+                source.data,
+                target_data,
+                target,
+                source_dim,
+                bounds_dim,
+                target_dim,
+                **transform_kwargs
+            )
+    else:
+        interpolated = conservative_interpolation(
+            source.data,
+            target_data,
+            target,
+            source_dim,
+            bounds_dim,
+            target_dim,
+            **transform_kwargs
+        )
+        xr.testing.assert_allclose(interpolated, expected.data)
+        # make sure that the extensive quantitiy is actually conserved
+        xr.testing.assert_allclose(interpolated.sum(), source.data.sum())
+
+
+"""High level tests"""
+
+
+def test_grid_transform(all_cases):
+    source, grid_kwargs, target, transform_kwargs, expected, error_flag = all_cases
+
+    axis = list(grid_kwargs["coords"].keys())[0]
+
+    grid = Grid(source, periodic=False, **grid_kwargs)
+
+    # the high level routines should be able to deal with all cases (no error flag exception like in the mid level)
+    transformed = grid.transform(source.data, axis, target, **transform_kwargs)
+    xr.testing.assert_allclose(transformed, expected.data)
+
+
+@pytest.mark.xfail(strict=True, raises=ValueError)
```

Comment:
Let's not use xfail. That's not the correct way to verify an error. Instead, within the test, use `with pytest.raises` with the `match` keyword inside the test.

> Using pytest.raises is likely to be better for cases where you are testing exceptions your own code is deliberately raising, whereas using @pytest.mark.xfail with a check function is probably better for something like documenting unfixed bugs (where the test describes what “should” happen) or bugs in dependencies.

https://docs.pytest.org/en/stable/assert.html#assertions-about-expected-exceptions

Reply:
Oops. Let me fix that real quick.

Reply:
What do you think about [these](https://github.com/xgcm/xgcm/blob/0e3b2213792ddcaaf6cfe763316e3e7e76f75023/xgcm/test/test_transform.py#L537-L538). That seems like an appropriate use of xfail?

Reply:
Yes, these seem like cases that we hope to implement eventually. Ideally we can remove those xfails and `error_flag` completely once this feature is fully done.